pkg/csi/cinder/controllerserver.go

```diff
@@ -439,123 +433,56 @@ func (cs *controllerServer) ListVolumes(ctx context.Context, req *csi.ListVolume
 		}
 	}
 
-	startIdx := 0
-	for _, cloudName := range cloudsNames {
-		if cloudName == currentCloudName {
-			break
-		}
-		startIdx++
-	}
-	for idx := startIdx; idx < len(cloudsNames); idx++ {
-		if maxEntries > 0 {
-			vlist, nextPageToken, err = cs.Clouds[cloudsNames[idx]].ListVolumes(ctx, maxEntries-len(cloudsVentries), startingToken)
-		} else {
-			vlist, nextPageToken, err = cs.Clouds[cloudsNames[idx]].ListVolumes(ctx, maxEntries, startingToken)
-		}
-		startingToken = nextPageToken
-		if err != nil {
-			klog.Errorf("Failed to ListVolumes: %v", err)
-			if cpoerrors.IsInvalidError(err) {
-				return nil, status.Errorf(codes.Aborted, "[ListVolumes] Invalid request: %v", err)
-			}
-			return nil, status.Errorf(codes.Internal, "ListVolumes failed with error %v", err)
-		}
-
-		ventries := cs.createVolumeEntries(vlist)
-		klog.V(4).Infof("ListVolumes: retrieved %d entries and %q next token from cloud %q", len(ventries), nextPageToken, cloudsNames[idx])
-
-		cloudsVentries = append(cloudsVentries, ventries...)
-
-		// Reach maxEntries setup nextToken with cloud identifier if needed
-		sendEmptyToken := false
-		if maxEntries > 0 && len(cloudsVentries) == maxEntries {
-			if nextPageToken == "" {
-				if idx+1 == len(cloudsNames) {
-					// no more entries and no more clouds
-					// send no token its finished
-					klog.V(4).Infof("ListVolumes: completed with %d entries and %q next token", len(cloudsVentries), "")
-					return &csi.ListVolumesResponse{
-						Entries:   cloudsVentries,
-						NextToken: "",
-					}, nil
-				} else {
-					// still clouds to process
-					// set token to next non empty cloud
-					i := 0
-					for i = idx + 1; i < len(cloudsNames); i++ {
-						vlistTmp, _, err := cs.Clouds[cloudsNames[i]].ListVolumes(ctx, 1, "")
-						if err != nil {
-							klog.Errorf("Failed to ListVolumes: %v", err)
-							if cpoerrors.IsInvalidError(err) {
-								return nil, status.Errorf(codes.Aborted, "[ListVolumes] Invalid request: %v", err)
-							}
-							return nil, status.Errorf(codes.Internal, "ListVolumes failed with error %v", err)
-						}
-						if len(vlistTmp) > 0 {
-							cloudsToken.CloudName = cloudsNames[i]
-							cloudsToken.isEmpty = false
-							break
-						}
-					}
-					if i == len(cloudsNames) {
-						sendEmptyToken = true
-					}
-				}
-			}
-			cloudsToken.CloudName = cloudsNames[idx]
-			cloudsToken.Token = nextPageToken
-			var data []byte
-			data, _ = json.Marshal(cloudsToken)
-			if sendEmptyToken {
-				data = []byte("")
-			}
-			klog.V(4).Infof("ListVolumes: completed with %d entries and %q next token", len(cloudsVentries), string(data))
-			return &csi.ListVolumesResponse{
-				Entries:   cloudsVentries,
-				NextToken: string(data),
-			}, nil
+	var volumeList []volumes.Volume
+	volumeList, token, err = cs.Clouds[cloudsNames[idx]].ListVolumes(ctx, maxEntries, token)
```

Comment:
we can simply use `cloudName` here
```suggestion
	volumeList, token, err = cs.Clouds[cloudName].ListVolumes(ctx, maxEntries, token)
```